### PR TITLE
Read the path and hashPath properties when reading .deps.json in hostfxr/policy

### PR DESCRIPTION
**Pull request by joelverhagen**
Since a "hashPath" and "path" property are now in the .deps.json file, this PR updates the native components in core-setup to prefer these values over manually constructing the paths.

Addresses the next point in https://github.com/NuGet/Home/issues/2522.

This is a reincarnation of https://github.com/dotnet/core-setup/pull/64.

/cc @gkhanna79 @schellap @eerhardt @livarcocc @piotrpMSFT

### CODE REVIEW

File: src/corehost/cli/deps_format.cpp [hunk 1]
```diff
@@ -43,6 +43,40 @@ void deps_json_t::reconcile_libraries_with_targets(
...
+        const auto& library_path_iter = properties.find(_X("path"));
+        if (library_path_iter == properties.end())
+        {
+            library_path = _X("");
+        }
+        else
+        {
+            library_path = library_path_iter->second.as_string();
+
+			if (_X('/') != DIR_SEPARATOR)
```
Comment (joelverhagen) [inline src/corehost/cli/deps_format.cpp]: Seems like I have some inconsistent indentation characters. Will fix.

Comment (joelverhagen) [inline src/corehost/cli/deps_format.cpp]: Fixed.

File: src/corehost/cli/deps_entry.cpp [hunk 2]
```diff
@@ -164,24 +172,49 @@ bool deps_entry_t::to_hash_matched_path(const pal::string_t& base, pal::string_t
...
-
-    // Build the hash file path str.
+    // Build the directory that contains the hash file.
     pal::string_t hash_file;
-    hash_file.reserve(base.length() + library_name.length() + library_version.length() + nupkg_filename.length() + 3);
-    hash_file.assign(base);
-    append_path(&hash_file, library_name.c_str());
-    append_path(&hash_file, library_version.c_str());
-    append_path(&hash_file, nupkg_filename.c_str());
+    if (library_path.empty())
```
Comment (eerhardt) [inline src/corehost/cli/deps_entry.cpp]: Is it possible to reuse the `deps_entry_t::to_full_path` method above?

File: src/corehost/cli/deps_format.cpp [hunk 3]
```diff
@@ -43,6 +43,40 @@ void deps_json_t::reconcile_libraries_with_targets(
...
+        }
+        else
+        {
+            library_hash_path = library_hash_path_iter->second.as_string();
+
+            if (_X('/') != DIR_SEPARATOR)
+            {
+                replace_char(&library_hash_path, _X('/'), DIR_SEPARATOR);
+            }
+        }
```
Comment (eerhardt) [inline src/corehost/cli/deps_format.cpp]: Can this chunk of code get extracted into a shared function between this and the library code above?  They look identical except for the string property name to look for.

Comment (joelverhagen) [inline src/corehost/cli/deps_format.cpp]: Yes, good idea!

File: src/corehost/cli/deps_entry.cpp (hunk 2 above)
Comment (joelverhagen) [inline src/corehost/cli/deps_entry.cpp]: `to_full_path` checks for file path existence on it's own, via `to_path`. `to_hash_matched_path` does it on it's own (`if (!fstream.good())`) with a specific trace message. I'd rather keep them separate to minimize risk. Does that seem reasonable?

File: src/corehost/cli/deps_format.cpp (hunk 3 above)
Comment (joelverhagen) [inline src/corehost/cli/deps_format.cpp]: Fixed.

File: src/corehost/cli/deps_entry.cpp (hunk 2 above)
Comment (eerhardt) [inline src/corehost/cli/deps_entry.cpp]: Yes, that seems totally fine.  Thanks for explaining.

**Comment by gkhanna79** [comment]
LGTM

**Comment by gkhanna79** [comment]
Please validate the scenarios (with and w/o the data) and confirm they work :)

**Comment by eerhardt** [comment]
:shipit:

**Pull request merged**